numpy/core/src/multiarray/datetime_strings.c

```diff
@@ -1592,3 +1593,302 @@ array_datetime_as_string(PyObject *NPY_UNUSED(self), PyObject *args,
 
     return NULL;
 }
+
+/*
+*  Format a string with datetime structure (internal helper)
+*/
+NPY_NO_EXPORT int
+strftime_from_datastruct(npy_datetimestruct dts, PyObject *timezone_obj,
+                         char *fmt, char *buff){
+    /* convert the fmt %f to string using dts */
+    char newfmt[80];
+    char ch, ch2;
+    char numstr[30];
+    int count, outcount;
+    int ndigits, ind0;
+    unsigned int ind;
+    int tzoffset;
+
+    /* Get the tzoffset from the timezone if provided */
+    tzoffset = -1;
+    if (timezone_obj != NULL) {
+        tzoffset = get_tzoffset_from_pytzinfo(timezone_obj, &dts);
+    }
+
+    if (tzoffset != -1){
+        add_minutes_to_datetimestruct(&dts, tzoffset);
+    }
+
+    outcount = 0;
+    for (count=0; fmt[count]!=0; ++count){
+        ch = fmt[count];
+        ndigits = -1;
+        if (ch == '%'){
+            /* check for %(%d)f , i.e. %8f  */
+            sscanf(&fmt[count], "%%%df", &ndigits);
+            if (ndigits > 0){
+                ch = fmt[++count];
+                ch = fmt[++count];
+                while ((ch != 'f') & (count < 70)){
```

Comment:
Here I think you mean "logical and", not "bitwise and":
```
while ((ch != 'f') && (count < 70)){
```
Also, what determines the limit of 70, and what should happen if it is exceeded?